crates/dojo-world/src/contracts/world.rs

```diff
@@ -1,334 +1,72 @@
-                selector: get_selector_from_name("grant_writer").unwrap(),
-            }])
-            .send()
-            .await
-            .map_err(WorldContractError::AccountError)
-    }
-
-    pub async fn register_models(
-        &self,
-        models: &[FieldElement],
-    ) -> Result<InvokeTransactionResult, AccountError<A::SignError>> {
-        let calls = models
-            .iter()
-            .map(|c| Call {
-                to: self.reader.address,
-                selector: selector!("register_model"),
-                calldata: vec![*c],
-            })
-            .collect::<Vec<_>>();
-
-        self.account.execute(calls).send().await
-    }
-
-    pub async fn deploy_contract(
-        &self,
-        salt: &FieldElement,
-        class_hash: &FieldElement,
-    ) -> Result<InvokeTransactionResult, AccountError<A::SignError>> {
-        self.account
-            .execute(vec![Call {
-                to: self.reader.address,
-                selector: selector!("deploy_contract"),
-                calldata: vec![*salt, *class_hash],
-            }])
-            .send()
-            .await
-    }
-
-    pub async fn executor(&self) -> Result<FieldElement, ContractReaderError> {
-        self.reader.executor().await
+#[cfg(doctest)]
+pub mod abigen {
+    pub mod world {
+        use cainome::rs::abigen;
+        abigen!(WorldContract, "src/contracts/abi/world.json");
     }
 
-    pub async fn base(&self) -> Result<FieldElement, ContractReaderError> {
-        self.reader.base().await
+    pub mod executor {
+        use cainome::rs::abigen;
+        abigen!(ExecutorContract, "src/contracts/abi/executor.json");
     }
-
-    pub async fn model(
-        &'a self,
-        name: &str,
-    ) -> Result<ModelRPCReader<'_, &'a A::Provider>, ModelError> {
-        self.reader.model(name).await
-    }
-}
-
-#[derive(Debug, thiserror::Error)]
-pub enum ContractReaderError {
-    #[error(transparent)]
-    ProviderError(#[from] ProviderError),
-    #[error(transparent)]
-    CairoShortStringToFeltError(#[from] CairoShortStringToFeltError),
-}
-
-pub struct WorldContractReader<P> {
-    provider: P,
-    block_id: BlockId,
-    address: FieldElement,
 }
 
 impl<P> WorldContractReader<P>
 where
-    P: Provider,
+    P: Provider + Sync + Send,
 {
-    pub fn new(address: FieldElement, provider: P) -> Self {
-        Self { address, provider, block_id: BlockId::Tag(BlockTag::Latest) }
-    }
-
-    pub fn with_block(self, block: BlockId) -> Self {
-        Self { block_id: block, ..self }
-    }
-
-    pub fn address(&self) -> FieldElement {
-        self.address
-    }
-
-    pub fn provider(&self) -> &P {
-        &self.provider
-    }
-
-    pub fn block_id(&self) -> BlockId {
-        self.block_id
+    pub async fn model_reader(&self, name: &str) -> Result<ModelRPCReader<'_, P>, ModelError> {
+        ModelRPCReader::new(name, self).await
     }
 }
 
-impl<P> WorldContractReader<P>
+impl<'a, P> WorldContractReader<P>
```

Comment:
```suggestion
impl<P> WorldContractReader<P>
```

what's this lifetime for ?

Reply:
I didn't removed it after some test, good catch. Interesting that the compiler is not complaining as `&self` is used, it should be implicit thus warned, no?